spec/models/reservation_spec.rb

```diff
@@ -1,725 +1,913 @@
-      r.save(validate: false)
-      r
+  describe '#checkin' do
+    let!(:handler) { FactoryGirl.build_stubbed(:checkout_person) }
+    context 'no procedures' do
+      let!(:res) { FactoryGirl.build_stubbed(:checked_out_reservation) }
+      it 'sets the checkin handler' do
+        expect { res.checkin(handler, {}, '') }.to \
+          change { res.checkin_handler }.from(nil).to(handler)
+      end
+      it 'sets the checked in time' do
+        # travel in order to freeze the time
+        travel(-1.days) do
+          expect { res.checkin(handler, {}, '') }.to \
+            change { res.checked_in }.from(nil).to(Time.zone.now)
+        end
+      end
+      it 'sets the status to returned' do
+        expect { res.checkin(handler, {}, '') }.to \
+          change { res.status }.from('checked_out').to('returned')
+      end
+      it 'updates the notes' do
+        # FIXME: might fall under overtesting (testing messages sent to self)
+        # but notes handling will be handled by a separate module in the future
+        # alternatively just test that the notes change?
+        # also applies to the next two tests
+        new_notes = ''
+        procedures = {}
+        allow(res).to receive(:make_notes)
+          .with('Checked in', new_notes, [], handler)
+        res.checkin(handler, procedures, new_notes)
+        expect(res).to have_received(:make_notes)
+          .with('Checked in', new_notes, [], handler)
+      end
+      it 'returns the reservation' do
+        expect(res.checkin(handler, {}, '')).to eq(res)
+      end
     end
-
-    it 'returns the correct late fee' do
-      expected = (Time.zone.today - reservation.due_date) *
-                 reservation.equipment_model.late_fee
-      expect(reservation.late_fee).to eq(expected)
+    context 'with completed procedures' do
+      it 'sends no procedures to the notes' do
+        # FIXME: this part of the method is really in need of refactoring
+        procedure = FactoryGirl.create(:checkin_procedure)
+        model = EquipmentModel.find(procedure.equipment_model.id)
+        res = FactoryGirl.build_stubbed(:checked_out_reservation,
+                                        equipment_model: model)
+        procedures = { procedure.id.to_s => '1' }
+        allow(res).to receive(:make_notes).with('Checked in', '', [], handler)
+        res.checkin(handler, procedures, '')
+        expect(res).to have_received(:make_notes)
+          .with('Checked in', '', [], handler)
+      end
     end
-
-    it 'returns 0 if not overdue' do
-      reservation.update_attributes(due_date: Time.zone.today + 1.day,
-                                    overdue: false)
-      expect(reservation.late_fee).to eq(0)
+    context 'with incomplete procedures' do
+      it 'sends the incomplete procedures to the notes' do
+        # FIXME: this part of the method is really in need of refactoring
+        procedure = FactoryGirl.create(:checkin_procedure)
```

Comment:
We don't, but I've been leaving the persisted data in other specs where mocking and stubbing was going to be more trouble than it was worth so I think leaving this is okay. 

Also, we'd stub so much here that it would be far too dependent on the current (bad) implementation